src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -653,9 +675,44 @@ public CompilationInfo(Compilation compilation, bool hasSuccessfullyLoaded)
                         }
                     }
 
-                    compilation = UpdateCompilationWithNewReferencesAndRecordAssemblySymbols(compilation, newReferences, metadataReferenceToProjectId);
+                    if (!Enumerable.SequenceEqual(compilation.ExternalReferences, newReferences))
+                    {
+                        compilation = compilation.WithReferences(newReferences);
+                    }
 
-                    this.WriteState(new FinalState(State.CreateValueSource(compilation, solution.Services), compilation, hasSuccessfullyLoaded), solution);
+                    // Now we run generators; if we don't have a generator driver at all, we must try create one
+                    if (generatorDriver.GeneratorDriver == null)
+                    {
+                        var generators = this.ProjectState.AnalyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                        var additionalTexts = this.ProjectState.AdditionalDocumentStates.Values.SelectAsArray(a => (AdditionalText)new AdditionalTextWithState(a));
+                        var compilationFactory = this.ProjectState.LanguageServices.GetRequiredService<ICompilationFactoryService>();
+
+                        generatorDriver = new TrackedGeneratorDriver(
+                            compilationFactory.CreateGeneratorDriver(
+                                this.ProjectState.ParseOptions!,
+                                generators,
+                                additionalTexts));
+                    }
+
+                    // We will finalize the compilation by adding full contents here.
+                    // PROTOTYPE: allow finalize compilation to incrementally update a prior version
+                    var compilationWithoutGeneratedFiles = compilation;
+
+                    if (generatorDriver.GeneratorDriver != null)
+                    {
+                        generatorDriver.GeneratorDriver.RunFullGeneration(compilation, out compilation, cancellationToken);
+                    }
+
+                    RecordAssemblySymbols(compilation, metadataReferenceToProjectId);
+
+                    this.WriteState(
```

Comment:
OK, just trying to understand this change, we basically made the after-generation compilation the only possible final compilation of state to make everything consistent?

Reply:
The CompilationTracker always moved Compilations through a bunch of states, finally reaching the "Final State". Any call to GetCompilationAsync() would simply push through to the FinalState and then use that one. The FinalState was also the only one that had P2P references wired up as you can see, so essentially that's strengthened here a bit.

The main change was that any given state is also allowed to provide a "good enough" compilation that it has at any point. That could be some partial compilation, or in the case of a FinalState it _is_ the final compilation. I'm now being more explicit and FinalState (if there are generators) now holds onto both the final compilation (post generator) and also the pre-generator compilation, where the latter is used for another later generation pass.

We'll probably have to tweak this further because the generator driver is also going to allow us to have incremental updating, which means we'll need to pull a post-generation compilation through some transformation states as well.